plugins/experimental/copy.cpp

```diff
@@ -1,54 +1,65 @@
-
-        auto operation(rsComm_t* comm, const json req) -> json
+    class copy : public parallel_filesystem_operation {
+        protected:
+        bool replace(std::string& str, const std::string& from, const std::string& to)
         {
-            try {
-                blackboard.set({{constants::status, states::running}});
-
-                fs::path source = get<std::string>("logical_path", req);
-                fs::path destination = get<std::string>("destination", req);
-                int thread_count = get<int>("thread_count", req, 4);
+            size_t start_pos = str.find(from);
+            if(start_pos == std::string::npos)
+                return false;
+            str.replace(start_pos, from.length(), to);
+            return true;
+        }
+
+        void process_object(rcComm_t& comm, const fs::path& path, const json& req) override
+        {
+            fs::path slp = get<std::string>("logical_path", req);
+            fs::path dlp = get<std::string>("destination_logical_path", req);
 
-                if(fsvr::is_collection(*comm, source)) {
-                    parallel_copy_collection(blackboard, {thread_count, source, destination});
-                }
-                else {
-                    fsvr::copy(*comm, source, destination);
+            if(req.contains("recursive")) {
+                auto replaced = path.string();
+                if(replace(replaced, slp, dlp)) {
+                    fscl::copy(comm, path, {replaced});
                 }
-
-                blackboard.update({{constants::status, states::complete}});
             }
-            catch(const irods::exception& e) {
-                rodsLog(LOG_ERROR, "%s", e.what());
-                blackboard.set({{constants::status,  states::failed},
-                                {constants::errors, {
-                                {constants::code,    e.code()},
-                                {constants::message, e.what()}}}});
+            else {
+                fscl::copy(comm, slp, dlp);
             }
+        } // process_object
 
-            return blackboard.get();
+        void collection_precondition(rcComm_t& comm, const json& req) override
+        {
+            fs::path slp = get<std::string>("logical_path", req);
+            fs::path dlp = get<std::string>("destination_logical_path", req);
+
+            if(!req.contains("recursive")) {
+                return;
+            }
 
-        } // operation
+            for(auto e : fscl::recursive_collection_iterator(comm, slp)) {
+                if(e.is_collection()) {
+                    auto replaced = e.path().string();
+
+                    if(replace(replaced, slp, dlp)) {
+                        auto dst = fs::path{replaced};
+                        try {
+                            if(!fscl::exists(comm, dst)) {
+                                fscl::create_collections(comm, dst);
+                            }
```

Comment:
you can remove the existence check here and call `create_collections` directly. It will check for existence before doing anything.